Criação da role do RKE

hazkall opened the pull request:
# Primeiro Role do RKE

<!-- Por favor descreva seu pull request aqui. -->

- [ x] Garanta que seu **topic/feature/bugfix branch** tenha uma branch nomeada e não a sua branch main esteja no PR
- [x ] Dê um titulo que expresse o objetivo do PR
- [x ] Associe seu PR a uma Issue criada no repositósito. Caso seja uma correção de linguagem ou pequenas correções, não é necessário
- [ x] Descreva o objetivo do PR
- [x ] Inclua links relevantes para a sua modificação/sugestão/correção
- [x ] Descreva um passo-a-passo para testar o seu PR

## Issue

<!-- Link da issue -->

https://github.com/mentoriaiac/iac-role-node-rke/issues/1

## Objetivo

Criado um role do ansible para criação de um node utilizando o RKE

## Referências

https://rancher.com/docs/rke/latest/en/os/

## Como testar

### Preparando o ambiente

#### Primeiro Passo:

Clone este repositório, ele é a base da role do nomad.
> git clone https://github.com/mentoriaiac/iac-role-node-rke


Ao final um diretório com a sua role será criado e você precisa agora entrar neste novo diretório. Para fazer isto use o comando a seguir:
> cd iac-role-node-rke


#### Segundo Passo
Crie um ambiente virtual para o python3. Como o ansible e o molecule são escritos em python, vamos precisar muito de uma estrutura de isolamento, como é o caso do modulo venv do python. Para criar um ambiente python isolado use o comando a seguir:

> python3 -m venv o-nome-da-sua-env


Você pode colocar o nome que você desejar para seu ambiente virtual, contudo existe uma convenção que estimula você usar este padrão: venv, .venv, env, .env

Uma vez criado seu ambiente virtual, precisamos inicializá-lo, para tal, execute o comando a seguir:

> source o-nome-da-sua-env/bin/activate


Se tudo der certo, seu pronpt será modificado e iniciará aparentemente assim:

> (o-nome-da-sua-env)$


Uma vez criado e inicializado o seu ambiente precisamos instalar todas as dependências necessárias para tudo acontecer certinho. As dependências desta role está gravada no arquivo ***requirements.txt***. Para instalar as dependências use o comando a seguir:

> (o-nome-da-sua-env)$ python3 -m pip install --upgrade pip

> (o-nome-da-sua-env)$ python3 -m pip install -r requirements.txt


#### Terceiro Passo
Para verificar se tudo aconteceu bem, você pode executar o comando a seguir:

> (o-nome-da-sua-env)$ molecule check


Ele ira executar um ciclo interiro de testes (dependency, cleanup, destroy, create, prepare, converge, check, cleanup, destroy)

#### Quarto Passo
Para realizar teste rápido após alguma modificação execute a seguinte sequência de comandos:

> (o-nome-da-sua-env)$ molecule create

> (o-nome-da-sua-env)$ molecule converge

> (o-nome-da-sua-env)$ molecule verify


Ao termino do teste, destrua o ambiente

> (o-nome-da-sua-env)$ molecule destroy


> (o-nome-da-sua-env)$ molecule check


Para realizar teste rápido após alguma modificação

> (o-nome-da-sua-env)$ molecule create

> (o-nome-da-sua-env)$ molecule converge

> (o-nome-da-sua-env)$ molecule verify


Ao termino do teste, destrua o ambiente

> (o-nome-da-sua-env)$ molecule destroy

<!--
Marque um `x` dentro de [ ] para os itens que você forneceu informação
Para modificar este template no seu repositório, basta criar o arquivo .github/pull_request_template.md nele.
-->

### Participantes do PR

 "@lipenodias - @monalizavitor - @marcelomansur - @victorfbraga - @ravcardoso - @guilhermegazzinelli  - @nerdtux"

somatorio commented on packer/provisioners/ansible/roles/kubeadm_docker_install/defaults/main.yml:
```diff
@@ -0,0 +1,14 @@
+usuario_default: mentoria
+
+package:
+  enabled: True
```
O ideal é que os arquivos terminem em uma linha em branco

Commit:
feat: Adicionado role para requisitos do RKE

Este commit adiciona o role ansible que faz o setup dos requisitos do RKE,
bem como adiciona os testes com molecule.

Também adicionei a tarefa no git actions.

Revisado por:
- Sebawebber
- Gabriel Iglesias
- Luiz Gustavo Rodrigues
- marcio_machado
- jhonatanrampin
Refs #1

Commit:
feat: adicionado README baseado no role do NOMAD

Revisado por:
- Sebawebber
- Gabriel Iglesias
- Luiz Gustavo Rodrigues
- marcio_machado
- jhonatanrampin
Refs #1

Commit:
ci: Adicionado teste de CI

Revisado por:
- Sebawebber
- Gabriel Iglesias
- Luiz Gustavo Rodrigues
- marcio_machado
- jhonatanrampin
Refs #1

Commit:
docs: Alterado nome dos autores da role

Revisado por:
- Sebawebber
- Gabriel Iglesias
- Luiz Gustavo Rodrigues
- marcio_machado
- jhonatanrampin
Refs #1

gomex commented on tasks/main.yml:
```diff
@@ -1,2 +1,100 @@
+
+- name: Criando usuário do RKE para criação do cluster
+  user:
+    name: k8s
+    groups: docker
+    generate_ssh_key: yes
+    shell: /bin/bash
+    state: present
```
Precisamos colocar uma chave publica pra permitir acesso remoto do rke.

sebawebber replied:
Faria sentido colocar uma lista de users e pegar as chaves publicas deles do github.

Ela fica disponível nesse modelo:

```
https://github.com/${GITHUB_USER}.keys
```

por exemplo, o meu: https://github.com/sebastianwebber.keys

hazkall replied:
Esse é um ponto que preciso de ajudar para entender a implementação.

sebawebber replied:
Pra implementar tem que usar o modulo `authorized_key`, conforme abaixo:

```yaml
- name: Set authorized key took from url
  authorized_key:
    user: charlie
    state: present
    key: https://github.com/charlie.keys
```

sebawebber replied:
btw @gomex, o que tu acha da gente adicionar um vault com a chave privada e usar essa chave no modulo que comentei acima?

talvez seja algo pra gente discutir no próximo encontro.

gomex replied:
Faz sentido @sebastianwebber, mas não precisamos subir no repositório. Podemos manter em um vault com acesso compartilhado, tal como lastpass ou bitwarden.

gomex replied:
Eu criei um usuário e coloquei a chave lá: https://github.com/mentoriaiacuser.keys

Eu criei tb o bitwarden pra uma conta compartilhada, pois o limite é de duas pessoas no plano free.

sebawebber replied:
ótima solução. vou colocar no TODO!

gomex commented on tasks/main.yml:
```diff
@@ -1,2 +1,100 @@
+  get_url:
+    url: "{{ rke_release }}"
+    dest: /usr/bin/rke
+    mode: +x
```
Não precisa do rke aqui, pois tudo é feito externamente.

gomex reviewed (changes requested):
Ainda não rodei pra testar de verdade, mas pelo que li, parece que estamos no caminho certo.

sebawebber commented on .ansible-lint:
```diff
@@ -7,3 +7,5 @@ offline: false
 parseable: true
 verbosity: 1
+warn_list:
+  - '106'
```
algum motivo em especial pra desabilitar esse warning?

hazkall replied:
Esse warning está ocorrendo devido ao nome da role colocado na mentoria
[E106] Role name iac-role-node-rke does not match ``^[a-z][a-z0-9_]+$`` pattern
Então como faz parte do projeto decidimos deixar na parte de ignore do lint

afonsoaugusto replied:
Vamos criar uma issue para renomear os repositórios de ansible com undescore (_) ao inves de hífen (-)

sebawebber replied:
https://github.com/mentoriaiac/iac-role-node-rke/issues/3

gomex replied:
Boa! Vamos deixar isso como débito técnico.

sebawebber commented on README.md:
```diff
@@ -1,33 +1,87 @@
+Crie um ambiente virtual para o python3. Como o ansible e o molecule são escritos em python, vamos precisar muito de uma estrutura de isolamento, como é o caso do modulo venv do python. Para criar um ambiente python isolado use o comando a seguir:
 
-BSD
+> python3 -m venv o-nome-da-sua-env
```
não seria mais simples apenas rodar o pipenv?

https://pipenv.pypa.io/en/latest/

gomex replied:
Podemos fazer sim, mas precisamos de uma proposta, pois até então todos usam venv

sebawebber commented on meta/main.yml:
```diff
@@ -14,25 +15,23 @@ galaxy_info:
   # - Apache-2.0
   # - CC-BY-4.0
-  license: license (GPL-2.0-or-later, MIT, etc)
+  license: GPL-2.0-or-later
```
algum motivo pra escolher essa licença em especial? talvez CC possa ser algo mais interessante pra o projeto. Opiniões?

hazkall replied:
Na verdade não foi alterado a licença no momento da realização do fork do projeto, não sei informar se ouve escolha das licenças. É um ponto interessante para revisar

gomex replied:
Vamos discutir a licença em outro momento.

sebawebber commented on files/daemon.json:
```diff
@@ -0,0 +1,8 @@
+    "exec-opts": ["native.cgroupdriver=systemd"],
+    "log-driver": "json-file",
+    "log-opts": {
+      "max-size": "100m"
```
por que 100m e não mais?

hazkall replied:
Nesse caso seguimos a documentação oficial do kubernetes para configuração do container runtime Docker com log para máximo de 100m.

<https://kubernetes.io/docs/setup/production-environment/container-runtimes/#docker>

gomex replied:
Acho que podemos manter 100m, mas não parece que esse seja o valor máximo que esse valor pode ter.

https://docs.docker.com/config/containers/logging/local/

"By default, the local driver preserves 100MB of log messages per container and uses automatic compression to reduce the size on disk"

sebastianwebber reviewed (changes requested):
Dei algumas sugestões, vou fazer testes na sequencia e com isso fazer novos comentários se necessário.

Commit:
feat: Adicionado suporte ao pipenv

BREAKING CHANGE: Este commit adiciona o pacote pipenv como dependencia. Com
isso a utilização do virtualenv ficou simplificada:

`pipenv shell` para abrir um terminal com o virtual env
`pipenv run $COMANDO` para executar diretamente dentro do virtual env

Também ajustado o CI para funcionar com pipenv.

Revisado por:
- Sebawebber
- Gabriel Iglesias
- Luiz Gustavo Rodrigues
- marcio_machado
- jhonatanrampin
Refs #1

sebawebber commented on .github/workflows/ci.yml:
```diff
@@ -1,25 +1,52 @@
+          python-version: ${{ matrix.python-version }}
+      - name: Install dependencies
+        run: |
+          python3 -m pip install --upgrade pip setuptools
```
É necessário ajustar para instalar o pipenv e também rodar o `pipenv install`

sebawebber commented on README.md:
```diff
@@ -1,33 +1,87 @@
+> (o-nome-da-sua-env)$
+
+
+Uma vez criado e inicializado o seu ambiente precisamos instalar todas as dependências necessárias para tudo acontecer certinho. As dependências desta role está gravada no arquivo ***requirements.txt***. Para instalar as dependências use o comando a seguir:
```
necessário ajustar o readme para colocar detalhes da instalação do pipenv.

Detalhes de como adicionar um pacote podem ser descritos com o link para a doc.

gomex replied:
Podemos fazer sim, mas precisamos de uma proposta, pois até então todos usam venv

gomex committed:
fix: Removeu rke e adicionar chave do k8s

Esse commit remove o binário do rke e ajusta os testes para garantir isso.

Adicionado a chave pública do usuário do deploy (k8s).

Revisado por:
 - Um monte de gente na live (Print nos comentários do PR)

 Ref #1

gomex committed:
feat: Adicionado editorconfig

Esse commit adiciona a configuração do editorconfig e formata os arquivos existentes.

Revisado por:
 - Um monte de gente na live (Print nos comentários do PR)

ref: #1

gomex committed:
doc: Adicionado changelog

Esse commit cria a primeira versão do CHANGELOG.

Revisado por:
 - Um monte de gente na live (Print nos comentários do PR)

ref: #1
